add Level Zero memory provider

bratpiorka opened the pull request:
This PR adds a Level Zero memory provider that supports #225 enhancement. TODO:
- [x] Make Level Zero provider a separate lib that dynamically loads Level Zero functions from ze_loader
- [x] Create a separate workflow for this with required deps (Level Zero) that will run on GPU
- [x] Add basic test with simple allocation/deallocation
- [x] Update README and docs

Please note that shared memory would be supported in separate PR

igchor commented:
@bratpiorka can you describe what's your plan regarding integration with UR? Do we want to have a duplicate provider in UMF for now or are you aiming to use this provider in UR? If it's the latter, there are more features you need to support: see the original implementation here https://github.com/oneapi-src/unified-runtime/blob/main/source/adapters/level_zero/usm.cpp#L170

igchor commented:
Also, the L0 provider should be a separate static library: see https://github.com/oneapi-src/unified-memory-framework/issues/225

bratpiorka commented:
@igchor thanks for the links. What do you think about creating a basic functionality first and then adding more features (in separate PRs) so it could finally replace the providers used in UR?

igchor commented:
> @igchor thanks for the links. What do you think about creating a basic functionality first and then adding more features (in separate PRs) so it could finally replace the providers used in UR?

Sure, that sounds reasonable, I just wanted to highlight that there might be some gaps with the current implementation.

igchor commented on include/umf/providers/provider_ze_memory.h:
```diff
@@ -0,0 +1,32 @@
+    ze_device_mem_alloc_desc_t device_mem_alloc_desc;
+} ze_memory_provider_params_t;
+
+umf_memory_provider_ops_t *umfZeMemoryProviderOps(void);
```
I think it should be called umfLevelZeroMemoryProvider (file as well). It would be more consistent with UR naming.

bratpiorka replied:
changed

igchor commented on src/libumf.map:
```diff
@@ -35,6 +35,7 @@ UMF_1.0 {
         umfMemoryProviderCreateFromMemspace;
         umfMemspaceCreateFromNumaArray;
         umfMemspaceDestroy;
+        umfZeMemoryProviderOps;
```
What about windows .map?

bratpiorka replied:
added

igchor commented on src/CMakeLists.txt:
```diff
@@ -58,6 +58,14 @@ if(UMF_BUILD_OS_MEMORY_PROVIDER)
+    set(UMF_SOURCES_LINUX ${UMF_SOURCES_LINUX}
+        provider/provider_ze_memory.c)
+    if(LINUX)
+        set(UMF_LIBS ${UMF_LIBS} ze_loader)
```
We should either:
- look for L0 using find_package() or similar
- use FetchContent to download L0 header (UR does that)

bratpiorka replied:
I changed this and ze_loader is loaded in runtime using dlopen

igchor commented on src/CMakeLists.txt:
```diff
@@ -58,6 +58,14 @@ if(UMF_BUILD_OS_MEMORY_PROVIDER)
+if(UMF_BUILD_ZE_MEMORY_PROVIDER)
+    set(UMF_SOURCES_LINUX ${UMF_SOURCES_LINUX}
+        provider/provider_ze_memory.c)
+    if(LINUX)
```
Why is this linux-specific? L0 also works on windows

bratpiorka replied:
Yes, but at the moment we don't have the internal infrastructure to test it

igchor commented on src/provider/provider_ze_memory.c:
```diff
@@ -0,0 +1,140 @@
+    ze_provider->device = ze_params->device;
+    ze_provider->device_mem_alloc_desc = ze_params->device_mem_alloc_desc;
+
+    assert(ze_provider->device_mem_alloc_desc.stype ==
```
Return INVALID_ARGUMENT here?

bratpiorka replied:
done

igchor commented on CMakeLists.txt:
```diff
@@ -8,6 +8,7 @@ project(unified-memory-framework VERSION 0.1.0 LANGUAGES C)
 # Build Options
 option(UMF_BUILD_SHARED_LIBRARY "Build UMF as shared library" OFF)
 option(UMF_BUILD_OS_MEMORY_PROVIDER "Build OS memory provider" ON)
+option(UMF_BUILD_ZE_MEMORY_PROVIDER "Build Level Zero memory provider" ON)
```
Default value (ON) does not match the README

bratpiorka replied:
done

vinser52 commented:
> @igchor thanks for the links. What do you think about creating a basic functionality first and then adding more features (in separate PRs) so it could finally replace the providers used in UR?

But what is the issue of migrating the whole UR implementation as is? Probably with some minor tweaks

vinser52 commented:
> Also, the L0 provider should be a separate static library: see #225

@igchor did you mean that L0 provider should be only static library or both: static and dynamic?

bratpiorka commented:
> > @igchor thanks for the links. What do you think about creating a basic functionality first and then adding more features (in separate PRs) so it could finally replace the providers used in UR?
> 
> But what is the issue of migrating the whole UR implementation as is? Probably with some minor tweaks

This version of the Level Zero provider is very simple and is intended to be used by oneCCL, but does not contain any oneCCL-specific code. On the other hand, the UR provider contains a lot of UR-specific code and is quite complex. Making it generic would require a lot of work. Also keep in mind that most of the contents of this PR are additional files needed by umf: workflows, cmake files, def/map etc. So even with this PR porting is still an option - it wouldn't change a lot of code.

igchor commented:
> > Also, the L0 provider should be a separate static library: see #225
> 
> @igchor did you mean that L0 provider should be only static library or both: static and dynamic?

I think we need the possibility to build it as a static library for UR. As for dynamic library, I don't know at this moment.

vinser52 commented:
> > > @igchor thanks for the links. What do you think about creating a basic functionality first and then adding more features (in separate PRs) so it could finally replace the providers used in UR?
> > 
> > 
> > But what is the issue of migrating the whole UR implementation as is? Probably with some minor tweaks
> 
> This version of the Level Zero provider is very simple and is intended to be used by oneCCL, but does not contain any oneCCL-specific code. On the other hand, the UR provider contains a lot of UR-specific code and is quite complex. Making it generic would require a lot of work. Also keep in mind that most of the contents of this PR are additional files needed by umf: workflows, cmake files, def/map etc. So even with this PR porting is still an option - it wouldn't change a lot of code.

My point is since UR already uses UMF and we agreed to move L0 provider impl from UR repo to UMF we should migrate it first (of course UR specific code should be cleaned up) and make sure we are not breaking UR flow. And then we can test if it fits for oneCCL needs.

vinser52 commented:
> > > Also, the L0 provider should be a separate static library: see #225
> > 
> > 
> > @igchor did you mean that L0 provider should be only static library or both: static and dynamic?
> 
> I think we need the possibility to build it as a static library for UR. As for dynamic library, I don't know at this moment.

Will it introduce any runtime dependencies to some L0 components? It is important in context of MPI, today they have no link dependency to L0 they try to load L0 at runtime (using dlopen) if I am correct. I will double-check.

bratpiorka commented:
> > > > @igchor thanks for the links. What do you think about creating a basic functionality first and then adding more features (in separate PRs) so it could finally replace the providers used in UR?
> > > 
> > > 
> > > But what is the issue of migrating the whole UR implementation as is? Probably with some minor tweaks
> > 
> > 
> > This version of the Level Zero provider is very simple and is intended to be used by oneCCL, but does not contain any oneCCL-specific code. On the other hand, the UR provider contains a lot of UR-specific code and is quite complex. Making it generic would require a lot of work. Also keep in mind that most of the contents of this PR are additional files needed by umf: workflows, cmake files, def/map etc. So even with this PR porting is still an option - it wouldn't change a lot of code.
> 
> My point is since UR already uses UMF and we agreed to move L0 provider impl from UR repo to UMF we should migrate it first (of course UR specific code should be cleaned up) and make sure we are not breaking UR flow. And then we can test if it fits for oneCCL needs.



> > > > Also, the L0 provider should be a separate static library: see #225
> > > 
> > > 
> > > @igchor did you mean that L0 provider should be only static library or both: static and dynamic?
> > 
> > 
> > I think we need the possibility to build it as a static library for UR. As for dynamic library, I don't know at this moment.
> 
> Will it introduce any runtime dependencies to some L0 components? It is important in context of MPI, today they have no link dependency to L0 they try to load L0 at runtime (using dlopen) if I am correct. I will double-check.

oneCCL also loads L0 at runtime

igchor commented:
> > > > Also, the L0 provider should be a separate static library: see #225
> > > 
> > > 
> > > @igchor did you mean that L0 provider should be only static library or both: static and dynamic?
> > 
> > 
> > I think we need the possibility to build it as a static library for UR. As for dynamic library, I don't know at this moment.
> 
> Will it introduce any runtime dependencies to some L0 components? It is important in context of MPI, today they have no link dependency to L0 they try to load L0 at runtime (using dlopen) if I am correct. I will double-check.

@vinser52  Okay, then in that case we'll need a dynamic version I guess. Can you add this info to the issue so that we don't miss it?

bratpiorka commented:
> > > > @igchor thanks for the links. What do you think about creating a basic functionality first and then adding more features (in separate PRs) so it could finally replace the providers used in UR?
> > > 
> > > 
> > > But what is the issue of migrating the whole UR implementation as is? Probably with some minor tweaks
> > 
> > 
> > This version of the Level Zero provider is very simple and is intended to be used by oneCCL, but does not contain any oneCCL-specific code. On the other hand, the UR provider contains a lot of UR-specific code and is quite complex. Making it generic would require a lot of work. Also keep in mind that most of the contents of this PR are additional files needed by umf: workflows, cmake files, def/map etc. So even with this PR porting is still an option - it wouldn't change a lot of code.
> 
> My point is since UR already uses UMF and we agreed to move L0 provider impl from UR repo to UMF we should migrate it first (of course UR specific code should be cleaned up) and make sure we are not breaking UR flow. And then we can test if it fits for oneCCL needs.

I looked into the code more deeply + consulted with @igchor and we could break the existing L0 provider in UR into two providers: a UR-specific top-level provider (implemented in UR) and a low-level generic L0 provider that just translates UMF provider API to ZE calls. With this approach, the provider implementation in this PR should be really simple and generic so it could be used by both UR and oneCCL

vinser52 commented:
> > > > > Also, the L0 provider should be a separate static library: see #225
> > > > 
> > > > 
> > > > @igchor did you mean that L0 provider should be only static library or both: static and dynamic?
> > > 
> > > 
> > > I think we need the possibility to build it as a static library for UR. As for dynamic library, I don't know at this moment.
> > 
> > 
> > Will it introduce any runtime dependencies to some L0 components? It is important in context of MPI, today they have no link dependency to L0 they try to load L0 at runtime (using dlopen) if I am correct. I will double-check.
> 
> @vinser52 Okay, then in that case we'll need a dynamic version I guess. Can you add this info to the issue so that we don't miss it?

Sure, I will update the issue. But, the dynamic version can be omitted if our static version will try to dlopen/dlsym `libze_loader.so` - In that case we are not making any hard dependency.

lukaszstolarczuk commented on .github/workflows/gpu.yml:
```diff
@@ -0,0 +1,71 @@
+            cd level_zero
+            mkdir build
+            cd build
+            cmake .. -DCMAKE_INSTALL_PREFIX=/usr
```
don't install in `/usr` if on self-hosted runner 😉 pls install locally, somewhere in workspace

lukaszstolarczuk commented on .github/workflows/gpu.yml:
```diff
@@ -0,0 +1,71 @@
+            mkdir build
+            cd build
+            cmake .. -DCMAKE_INSTALL_PREFIX=/usr
+            sudo cmake --build . --config Release --target install
```
sudo won't be needed

pbalcer commented:
> > 
> 
> Perhaps, but I think UR might prefer to have a static library that doesn't perform any dlopens - at least for the case where all dependencies are linked statically. @pbalcer is that true? or is there no difference for UR?

UR would require two variants of this provider: one that is fully dynamically linked that uses the exact same L0 symbols as the adapter (so not dlopen'ed), and one that uses static linking and then uses symbols that from the library that the adapter already dlopen'ed (so we'd probably need to dlopen the exact same library path).

bratpiorka commented:
> Let's discus it on our sync on Monday. I doubt if we need dlopen at all. maybe just dlsym is enough. My thoughts are the following:
> 
> * L0 memory provider is used only when client already uses L0.
> * So L0 should already be loaded into the process.
> * UMF should just find the corresponding  L0 symbols that are already loaded in the process somehow. It should be possible with `dlsym` and `RTLD_DEFAULT` (a special pseudo-handle)
> 
> With such schema, clients will use L0 in the same way like today (explicit linking, dlopen, etc.). UMF will not have dependency to `libze_loader`. L0 memory provider implementation will try to find corresponding symbols at runtime.

@vinser52 I made changes that follow this algorithm: first I try to get symbols using NULL handle and only if this fails I try to load libze_loader.so

bratpiorka commented:
@pbalcer I will resolve all UR-related issues in a separate PR when I also try to do a PR in UR repo that integrates with this provider

igchor commented:
> > > > 
> > > 
> > > 
> > > Perhaps, but I think UR might prefer to have a static library that doesn't perform any dlopens - at least for the case where all dependencies are linked statically. @pbalcer is that true? or is there no difference for UR?
> > 
> > 
> > UR would require two variants of this provider: one that is fully dynamically linked that uses the exact same L0 symbols as the adapter (so not dlopen'ed), and one that uses static linking and then uses symbols that from the library that the adapter already dlopen'ed (so we'd probably need to dlopen the exact same library path).
> 
> Let's discus it on our sync on Monday. I doubt if we need dlopen at all. maybe just dlsym is enough. My thoughts are the following:
> 
> * L0 memory provider is used only when client already uses L0.
> * So L0 should already be loaded into the process.
> * UMF should just find the corresponding  L0 symbols that are already loaded in the process somehow. It should be possible with `dlsym` and `RTLD_DEFAULT` (a special pseudo-handle)
> 
> With such schema, clients will use L0 in the same way like today (explicit linking, dlopen, etc.). UMF will not have dependency to `libze_loader`. L0 memory provider implementation will try to find corresponding symbols at runtime.

@vinser52 @pbalcer  But keep in mind that what Rafal implemented is using dlopen/dlsym on level zero **loader**. The loader itself will use dlopen to load appropriate drivers (from what I understand). So I don't think we need to worry about what symbols are being used, do we? Btw, assuming there would be a static version of L0 loader (could we ask L0 team?), could we just link with that and rely on L0 discovery instead of doing any dlopen/dlsym ourselves?

igchor commented on test/CMakeLists.txt:
```diff
@@ -124,6 +125,12 @@ if(UMF_BUILD_OS_MEMORY_PROVIDER AND LINUX) # OS-specific functions are implement
                  LIBS umf_utils numa)
 endif()
 
+if(UMF_BUILD_LIBUMF_PROVIDER_LEVEL_ZERO AND LINUX) # TODO
```
Why TODO? Is there anything missing to support it on windows?

bratpiorka replied:
At the moment we don't have the internal infrastructure to test it. I will try to include Windows ASAP in separate PRs

igchor commented on test/providers/provider_level_zero.cpp:
```diff
@@ -0,0 +1,210 @@
+    const char *message;
+    int32_t error = 0;
+    umfMemoryProviderGetLastNativeError(provider, &message, &error);
+    ASSERT_EQ(error, 0);
```
I don't think you should test for that. The return value of umfMemoryProviderGetLastNativeError is unspecified if the previous call was successful.

bratpiorka replied:
removed

igchor commented on test/providers/provider_level_zero.cpp:
```diff
@@ -0,0 +1,210 @@
+        ASSERT_NE(hContext, nullptr);
+    }
+
+    void TearDown() override { test::TearDown(); }
```
shouldn't you destroy the context, etc. here?

bratpiorka replied:
thanks, added

bratpiorka commented on README.md:
```diff
@@ -31,6 +31,10 @@ For development and contributions:
 - C++ compiler with C++17 support
 
+For Level Zero memory provider and its tests:
+- Level Zero headers and libraries
```
I've added this requirement one line down - if you have a better idea how to write it, let me know :)

bratpiorka commented on src/provider/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# libumf_provider_level_zero
+if(UMF_BUILD_LIBUMF_PROVIDER_LEVEL_ZERO)
```
Yes, but the Level Zero memory provider is a separate library, while the OS memory provider is not (there is only an option to enable/disable it by adding its sources to libumf). If we decide to make it a separate, static library (linked by default), it should be done in a separate PR.

bratpiorka replied:
Right now I use ze_api.h header but do you suggest removing it completely? I'm not sure if I should re-define Level Zero structures inside UMF

vinser52 replied:
but how it relates to the fact that L0 provider is a separate static library?

igchor reviewed (approved):
I'd still like to discuss the linking (whether and how we use dlopen) on our sync but the rest looks good.

bratpiorka commented:
@igchor @vinser52 @pbalcer - please review. I removed dlopen() from the Level Zero provider at all. The test I added to the CI loads libze_loader dynamically using dlopen(RTLD_GLOBAL) and the Level Zero provider reuses these symbols without any problems.

bratpiorka commented on src/libumf.map:
```diff
@@ -35,6 +35,7 @@ UMF_1.0 {
         umfMemoryProviderCreateFromMemspace;
         umfMemspaceCreateFromNumaArray;
         umfMemspaceDestroy;
+        umfLevelZeroMemoryProviderOps;
```
done

bratpiorka commented on .github/workflows/gpu.yml:
```diff
@@ -0,0 +1,47 @@
+
+      - name: Run tests
+        working-directory: ${{github.workspace}}/build
+        run: |
```
done

igchor commented on include/umf/providers/provider_level_zero.h:
```diff
@@ -0,0 +1,34 @@
+#ifndef UMF_PROVIDER_LEVEL_ZERO_H
+#define UMF_PROVIDER_LEVEL_ZERO_H
+
+#include <level_zero/ze_api.h>
```
I don't think we can include this header here. Our customers might not have the headers installed, if they use dlopen, right? @vinser52?

bratpiorka replied:
but how they could pass ze_context_handle_t or ze_device_handle_t to the level zero provider params?

igchor replied:
They might have a copy of the headers somewhere, but not necessarily in <level_zero/ze_api.h> I think.

bratpiorka replied:
or I can check the #define and include the file only when the user hadn't done it yet + print the messafe

igchor replied:
We can also just accept 'void*' for each param - I don't think there is a good solution for that if we want to use dlopen.

bratpiorka replied:
Ok I removed this header from the interface

igchor commented:
> @igchor @vinser52 @pbalcer - please review. I removed dlopen() from the Level Zero provider at all. The test I added to the CI loads libze_loader dynamically using dlopen(RTLD_GLOBAL) and the Level Zero provider reuses these symbols without any problems.

Can we also have a similar test for when the application links with Level Zero normally (not using dlopen)? I.e. just have a second test that just assign all those functions pointers in libze_ops to &zeInit, &zeDriverGet, etc?

lukaszstolarczuk reviewed (commented):
👍

bratpiorka commented:
> Can we also have a similar test for when the application links with Level Zero normally (not using dlopen)? I.e. just have a second test that just assign all those functions pointers in libze_ops to &zeInit, &zeDriverGet, etc?

@igchor done

vinser52 commented on CMakeLists.txt:
```diff
@@ -8,6 +8,7 @@ project(unified-memory-framework VERSION 0.1.0 LANGUAGES C)
 # Build Options
 option(UMF_BUILD_SHARED_LIBRARY "Build UMF as shared library" OFF)
 option(UMF_BUILD_OS_MEMORY_PROVIDER "Build OS memory provider" ON)
+option(UMF_BUILD_LIBUMF_PROVIDER_LEVEL_ZERO "Build Level Zero memory provider" OFF)
```
if L0 provider implementation has no build-time dependencies why not ON it by default?

bratpiorka replied:
no, ze_api.h is still used in src/provider/provider_level_zero.c

vinser52 replied:
Ok, there is a build-time dependency. Another question regarding distribution, since L0 provider implementation has no runtime dependency to L0 driver we can make L0 provider implementation to be part of core UMF library.

bratpiorka replied:
I will make this option enabled by default (ON) + check for build deps in cmake + message to user

vinser52 commented on src/provider/provider_level_zero.c:
```diff
@@ -0,0 +1,240 @@
+    ze_provider->context = ze_params->level_zero_context_handle;
+    ze_provider->device = ze_params->level_zero_device_handle;
+    ze_provider->memory_type = ze_params->memory_type;
+    ze_provider->host_mem_alloc_desc = (ze_host_mem_alloc_desc_t){
+        ZE_STRUCTURE_TYPE_HOST_MEM_ALLOC_DESC,
+        NULL, // pointer to an extension-specific structure
+        ze_params->level_zero_host_mem_alloc_flags};
+    ze_provider->device_mem_alloc_desc = (ze_device_mem_alloc_desc_t){
+        ZE_STRUCTURE_TYPE_DEVICE_MEM_ALLOC_DESC,
+        NULL, // pointer to an extension-specific structure
+        ze_params->level_zero_device_mem_alloc_flags,
+        ze_params->level_zero_device_local_mem_ordinal};
```
I do not understand where it is used and why it is needed?

bratpiorka replied:
you have to pass ze_device_mem_alloc_desc_t struct as an argument to zeMemAllocDevice

vinser52 replied:
right, but we can create it on stack inside `ze_memory_provider_alloc` function. Furthermore, `device_mem_alloc_desc` is needed only when `ze_provider->memory_type` equals to the device memory - it is another argument not to make it part of the `ze_params` object.

vinser52 commented on src/provider/provider_level_zero.c:
```diff
@@ -0,0 +1,240 @@
+    assert(resultPtr);
+
+    ze_memory_provider_t *ze_provider = (ze_memory_provider_t *)provider;
+    ze_result_t ze_result = g_ze_ops.zeMemAllocDevice(
```
Based on `ze_provider->memory_type` we should use `zeMemAllocDevice`, `zeMemAllocHost` or `zeMemAllocShared`.

bratpiorka replied:
done

vinser52 commented on src/provider/provider_level_zero.c:
```diff
@@ -0,0 +1,240 @@
+    (void)ptr;
+
+    // TODO
+    *pageSize = 1024;
```
as I know it is 64K

bratpiorka commented on src/provider/provider_level_zero.c:
```diff
@@ -0,0 +1,240 @@
+} ze_memory_provider_t;
+
+typedef struct ze_ops_t {
+    ze_result_t (*zeMemAllocDevice)(ze_context_handle_t,
```
I planned to add shared and host in the next PR which should contain all modes required by UR

bratpiorka replied:
added

bratpiorka commented on README.md:
```diff
@@ -123,6 +128,18 @@ A memory provider that provides memory from an operating system.
+2) Linux OS
+3) The `UMF_BUILD_LIBUMF_PROVIDER_LEVEL_ZERO` option turned `ON`
+4) Required packages:
+   - TODO
```
is this a package? I rather thought of libze1 but I'm not sure

vinser52 replied:
Ok, I think I got the point. To build L0 provider we need `level-zero-dev` (on Ubuntu, maybe on RedHat it named differently) package. But for distribution, we do not need such dependency, right?

bratpiorka replied:
done

bratpiorka commented on README.md:
```diff
@@ -123,6 +128,18 @@ A memory provider that provides memory from an operating system.
 
+#### Level Zero memory provider (Linux-only yet)
+
+A memory provider that provides memory from a graphics device.
```
done

bratpiorka commented:
@vinser52 I have resolved the discussions started by you, as we agreed

ldorau commented on .github/workflows/pr_push.yml:
```diff
@@ -113,3 +115,6 @@ jobs:
     uses: ./.github/workflows/proxy_lib.yml
+  GPU:
+    needs: [Build]
+    uses: ./.github/workflows/gpu.yml
```
We may add badges for new workflows later ...

bratpiorka replied:
added

ldorau reviewed (changes requested):
It does not build on WSL Ubuntu:
```
/home/ldorau/work/unified-memory-framework/src/provider/provider_level_zero.c: In function ‘init_ze_global_state’:
/home/ldorau/work/unified-memory-framework/src/provider/provider_level_zero.c:48:48: error: ‘RTLD_DEFAULT’ undeclared (first use in this function)
   48 |     *(void **)&g_ze_ops.zeMemAllocHost = dlsym(RTLD_DEFAULT, "zeMemAllocHost");
      |                                                ^~~~~~~~~~~~
/home/ldorau/work/unified-memory-framework/src/provider/provider_level_zero.c:48:48: note: each undeclared identifier is reported only once for each function it appears in
make[2]: *** [src/CMakeFiles/umf.dir/build.make:328: src/CMakeFiles/umf.dir/provider/provider_level_zero.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1153: src/CMakeFiles/umf.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

bratpiorka commented:
@ldorau please check now. BTW we definitely should add WSL to our CI
